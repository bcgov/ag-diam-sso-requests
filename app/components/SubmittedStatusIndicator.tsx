import { ProgressBar } from 'react-bootstrap';
import Title from 'components/SHeader3';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faInfoCircle, faCheckCircle, faTimesCircle, faSpinner } from '@fortawesome/free-solid-svg-icons';
import Link from '@button-inc/bcgov-theme/Link';
import styled from 'styled-components';
import { LINK_COLOR, SECONDARY_BLUE } from 'styles/theme';
import HelpText from 'components/HelpText';
import { Request } from 'interfaces/Request';
import getConfig from 'next/config';
import StatusList from 'components/StatusList';

const { publicRuntimeConfig = {} } = getConfig() || {};
const { app_env } = publicRuntimeConfig;

interface Props {
  selectedRequest: Request;
}

const StyledIcon = styled(FontAwesomeIcon)`
  color: ${SECONDARY_BLUE};
  padding-right: 5px;
`;

const SubTitle = styled(Title)`
  margin-top: 20px;
  border-bottom: none;
`;

const SLink = styled.a`
  color: ${LINK_COLOR};
`;

const SProgressBar = styled(ProgressBar)`
  margin-bottom: 10px;
`;

const getPercent = (status?: string) => {
  switch (status) {
    case 'submitted':
      return 0;
    case 'pr':
      return 33;
    case 'planned':
      return 66;
    default:
      return 100;
  }
};

const getStatusMessage = (status?: string) => {
  switch (status) {
    case 'submitted':
      return 'Process request submitted...';
    case 'pr':
      return 'Pull request created...';
    case 'planned':
      return 'Terraform plan succeeded...';
    case 'prFailed':
    case 'planFailed':
    case 'applyFailed':
      return 'An error has occurred';
    default:
      return '';
  }
};

const getStatusFailure = (status?: string) => {
  switch (status) {
    case 'prFailed':
    case 'planFailed':
    case 'applyFailed':
      return true;
    default:
      return false;
  }
};

const getStatusStatusCode = (status?: string) => {
  switch (status) {
    case 'submitted':
      return 0;
    case 'pr':
      return 1;
    case 'prFailed':
      return 2;
    case 'planned':
      return 3;
    case 'planFailed':
      return 4;
    case 'applied':
      return 5;
    case 'applyFailed':
      return 6;
    default:
      return 0;
  }
};

export default function SubmittedStatusIndicator({ selectedRequest }: Props) {
  const { status, prNumber, updatedAt } = selectedRequest;

  const hasError = getStatusFailure(status);
  const statusMessage = getStatusMessage(status);
  const formattedUpdatedAt = new Date(updatedAt || '').toLocaleString();

  // Step 1.
  const statusItems = [
    <li key="1">
      Process request submitted
      <FontAwesomeIcon icon={faCheckCircle} color="#00C45B" />
    </li>,
  ];

  const code = getStatusStatusCode(status);

  // Step 2.
  if (code < 1) {
    statusItems.push(
      <li key="2">
        Pull request created
        <FontAwesomeIcon icon={faSpinner} color="#B2B2B2" spin />
      </li>,
    );
  } else {
    if (code === 2) {
      statusItems.push(
        <li key="2">
          Pull request creation failed
          <FontAwesomeIcon icon={faTimesCircle} color="#FF0000" />
        </li>,
      );
    } else {
      const prLink =
        app_env === 'production'
          ? `https://github.com/bcgov/sso-terraform/pull/${prNumber}`
          : `https://github.com/bcgov/sso-terraform-dev/pull/${prNumber}`;

      statusItems.push(
        <li key="2">
          Pull request created (
          <Link external href={prLink}>
            link
          </Link>
          )
          <FontAwesomeIcon icon={faCheckCircle} color="#00C45B" />
        </li>,
      );
    }
  }

  // Step 3.
  if (code < 3) {
    statusItems.push(
      <li key="3">
        Terraform plan succeeded
        <FontAwesomeIcon icon={faSpinner} color="#B2B2B2" spin />
      </li>,
    );
  } else {
    if (code === 4) {
      statusItems.push(
        <li key="3">
          Terraform plan failed
          <FontAwesomeIcon icon={faTimesCircle} color="#FF0000" />
        </li>,
      );
    } else {
      statusItems.push(
        <li key="3">
          Terraform plan succeeded
          <FontAwesomeIcon icon={faCheckCircle} color="#00C45B" />
        </li>,
      );
    }
  }

  // Step 4.
  if (code < 5) {
    statusItems.push(
      <li key="4">
        Request processed
        <FontAwesomeIcon icon={faSpinner} color="#B2B2B2" spin />
      </li>,
    );
  } else {
    if (code === 6) {
      statusItems.push(
        <li key="4">
          Request failed
          <FontAwesomeIcon icon={faTimesCircle} color="#FF0000" />
        </li>,
      );
    } else {
      statusItems.push(
        <li key="4">
          Request processed
          <FontAwesomeIcon icon={faCheckCircle} color="#00C45B" />
        </li>,
      );
    }
  }

  return (
    <>
      <Title>
        <StyledIcon icon={faInfoCircle} />
        We are working on your request - Expected processing time is 20 mins
      </Title>
      <SubTitle>{statusMessage}</SubTitle>
      <SProgressBar now={getPercent(status)} animated variant={hasError ? 'danger' : undefined} />
      <HelpText>Last updated at {formattedUpdatedAt}</HelpText>
      <StatusList>{statusItems}</StatusList>
      <br />
      <p>
        If there is an error or the process takes longer than 20 mins then,
        please contact our SSO support team by{' '}
        <SLink href="https://chat.developer.gov.bc.ca/channel/sso" target="_blank" title="Rocket Chat">
          Rocket.Chat
        </SLink>{' '}
        or{' '}
        <SLink href="mailto:bcgov.sso@gov.bc.ca" title="Pathfinder SSO" target="blank">
          Email us
        </SLink>{' '}
      </p>
    </>
  );
}
