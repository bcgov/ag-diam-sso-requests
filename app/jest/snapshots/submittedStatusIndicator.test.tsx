import SubmittedStatusIndicator from 'components/SubmittedStatusIndicator';
import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

it('should match the snapshot', () => {
  let { asFragment } = render(<SubmittedStatusIndicator status="pr" updatedAt="" />);
  expect(asFragment()).toMatchSnapshot();

  ({ asFragment } = render(<SubmittedStatusIndicator status="prFailed" updatedAt="" />));
  expect(asFragment()).toMatchSnapshot();
});
