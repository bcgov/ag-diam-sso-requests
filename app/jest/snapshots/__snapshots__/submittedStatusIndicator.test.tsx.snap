// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match the snapshot 1`] = `
<DocumentFragment>
  <h3
    class="sc-bdnxRM eCWcRC"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-info-circle fa-w-16 sc-dlnjwi hEbaYC"
      data-icon="info-circle"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
        fill="currentColor"
      />
    </svg>
    We are working on your request - Expected processing time is 20 mins
  </h3>
  <h3
    class="sc-bdnxRM sc-hKFxyN eCWcRC jIPEne"
  >
    Pull request created...
  </h3>
  <div
    class="sc-jSFjdj hObHgj progress"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="33"
      class="progress-bar progress-bar-animated progress-bar-striped"
      role="progressbar"
      style="width: 33%;"
    />
  </div>
  <p
    class="sc-gtsrHT dHnfDJ"
  >
    Last updated at Invalid Date
  </p>
</DocumentFragment>
`;

exports[`should match the snapshot 2`] = `
<DocumentFragment>
  <h3
    class="sc-bdnxRM eCWcRC"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-info-circle fa-w-16 sc-dlnjwi hEbaYC"
      data-icon="info-circle"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 512 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 8C119.043 8 8 119.083 8 256c0 136.997 111.043 248 248 248s248-111.003 248-248C504 119.083 392.957 8 256 8zm0 110c23.196 0 42 18.804 42 42s-18.804 42-42 42-42-18.804-42-42 18.804-42 42-42zm56 254c0 6.627-5.373 12-12 12h-88c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h12v-64h-12c-6.627 0-12-5.373-12-12v-24c0-6.627 5.373-12 12-12h64c6.627 0 12 5.373 12 12v100h12c6.627 0 12 5.373 12 12v24z"
        fill="currentColor"
      />
    </svg>
    We are working on your request - Expected processing time is 20 mins
  </h3>
  <h3
    class="sc-bdnxRM sc-hKFxyN eCWcRC jIPEne"
  >
    An error has occurred
  </h3>
  <div
    class="sc-jSFjdj hObHgj progress"
  >
    <div
      aria-valuemax="100"
      aria-valuemin="0"
      aria-valuenow="100"
      class="progress-bar bg-danger progress-bar-animated progress-bar-striped"
      role="progressbar"
      style="width: 100%;"
    />
  </div>
  <p
    class="sc-gtsrHT dHnfDJ"
  >
    Last updated at Invalid Date
  </p>
  <p>
    Please contact our SSO support team by
    <a
      class="sc-eCApnc jSSwhS"
      href="https://chat.developer.gov.bc.ca/channel/sso"
      target="_blank"
      title="Rocket Chat"
    >
      Rocket.Chat
    </a>
     or
    <a
      class="sc-eCApnc jSSwhS"
      href="mailto:zorin.samji@gov.bc.ca"
      target="blank"
      title="Pathfinder SSO"
    >
      Email us
    </a>

  </p>
</DocumentFragment>
`;
